src/array/cuda/spmm.cu

```diff
@@ -400,6 +501,109 @@ void SpMMCsr(const std::string& op, const std::string& reduce,
   }
 }
 
+/*!
+ * \brief CUDA implementation of g-SpMM on Csr format.
+ * \note use cusparse if the reduce operator is `sum` and there is
+ *       no broadcast, use dgl's kernel in other cases.
+ */
+template <int XPU, typename IdType, int bits>
+void SpMMCsrHetero(const std::string& op, const std::string& reduce,
+             const BcastOff& bcast,
+             const std::vector<CSRMatrix>& vec_csr,
+             std::vector<NDArray> vec_ufeat,
+             std::vector<NDArray> vec_efeat,
+             std::vector<NDArray> vec_out,
+             std::vector<NDArray> out_aux,
+             const std::vector<dgl_type_t> ufeat_eid,
+             const std::vector<dgl_type_t> out_eid) {
+  int64_t feat_len = bcast.out_len;
+  bool is_scalar_efeat = vec_efeat.size() != 0; //efeat.NumElements() == csr.indices->shape[0];
+  bool use_efeat = op != "copy_lhs";
+  const DLContext& ctx = DLContext{kDLGPU, 0};
+  int maxstrm = 16;
+  cudaStream_t stream[maxstrm];
+  for (int i = 0; i < 16; i++)
+    cudaStreamCreate(&(stream[i]));
+  for (dgl_type_t etype = 0; etype < ufeat_eid.size(); ++etype) {
+    const dgl_type_t src_id = ufeat_eid[etype];
+    const dgl_type_t dst_id = out_eid[etype];
+    CSRMatrix csr = vec_csr[etype];
+    if (reduce == "sum") {
+      SWITCH_BITS(bits, DType, {
+        /* Call  SpMM for each relation type */
+        if (op == "copy_lhs" && cusparse_available<bits, IdType>()) {  // cusparse
+          int64_t x_length = 1;
+          NDArray nd_ufeat = vec_ufeat[ufeat_eid[0]];
+          for (int i = 1; i < nd_ufeat->ndim; ++i){
+            x_length *= nd_ufeat->shape[i]; //TODO:: All ufeat shares same shape?
+          }
+          cusparse::CusparseCsrmm2Hetero<DType, IdType>(
+              ctx, csr,
+              static_cast<DType*>(vec_ufeat[src_id]->data),
+              nullptr,
+              static_cast<DType*>(vec_out[dst_id]->data),
+              x_length,
+              stream[etype % maxstrm]);
```

Comment:
This seems to be the only operation here that's running on a non-default stream. Shouldn't everything in here run on stream[etype % maxstrm]?